Tests/ShamTests/Tests/MockServiceTests.swift

```diff
@@ -103,6 +107,48 @@ final class MockServiceTests: XCTestCase {
         XCTAssertFalse(hasData)
     }
     
+    func testReturnsStubWithMostEqualParameters() throws {
+        // GIVEN: I stub a URL with only one query.
+        self.stub(self.fullURL, with: .encodable(self.mockData))
+        
+        // THEN: The values should return correctly even if the URL that is requested has multiple
+        // query parameters.
+        self.request(url: self.manyQueryParamsURL, data: self.mockData)
+    }
+    
+    func testReturnsStubWithMostEqualParametersBasedOnBestMatch() throws {
+        // GIVEN: Two stub URLs have the same base base URL and path.
+        let fullURL = try XCTUnwrap(URL(string: self.fullURL))
+        let queryLessURL = try XCTUnwrap(URL(string: self.querylessURL))
+        XCTAssertEqual(fullURL.baseURL, queryLessURL.baseURL)
+        XCTAssertEqual(fullURL.path, queryLessURL.path)
+        
+        // GIVEN: I stub similar URLs with different query params.
+        self.stub(self.querylessURL, with: .encodable(self.mockData))
+        self.stub(self.fullURL, with: .encodable(self.secondaryMockData))
+        
+        // THEN: The values should return correctly.
+        self.request(url: self.querylessURL, data: self.mockData)
+        self.request(url: self.fullURL, data: self.secondaryMockData)
+        
+        self.request(url: self.manyQueryParamsURL, data: self.secondaryMockData)
+    }
+    
+    func testReturnsStubWithQueryParametersEvenIfTheRequestDoesNotHaveQueryParameters() throws {
+        // GIVEN: Two stub URLs have the same base base URL and path.
+        let fullURL = try XCTUnwrap(URL(string: self.fullURL))
+        let queryLessURL = try XCTUnwrap(URL(string: self.querylessURL))
+        XCTAssertEqual(fullURL.baseURL, queryLessURL.baseURL)
+        XCTAssertEqual(fullURL.path, queryLessURL.path)
+        
+        // GIVEN: I stub a URL with query params.
+        self.stub(self.fullURL, with: .encodable(self.mockData))
+        
+        // THEN: The value should return correctly, even if the request does not have query
+        // params.
+        self.request(url: self.querylessURL, data: self.mockData)
```

Comment:
I'm a little worried that while this solution may aid some use cases, it may be harmful in others.

As an example, say we have the following in our API/app behavior:
- Our API endpoint `foo` requires a query parameter of `foo` to produce a valid response
- If the query parameter `foo` is not present, an error message should be displayed.

So:
- `"https://foo.com/foo?foo=bar"` -> Should succeed and take user to next page in the app
- `"https://foo.com/foo"` -> Should fail and display an error alert to the user

In my UI test, I've stubbed `"https://foo.com/foo?foo=bar"` with a 200 response. There is a bug in my app and we're not actually sending the query parameter, so the request going out looks like `"https://foo.com/foo`, which should produce an error alert but instead grabs the success response and moves the user forward.

This is definitely an interesting solution to the problem we were facing, but I think it could be seen as as undesired behavior in certain circumstances and the user cannot turn it off if they don't like it.

I'm curious if we could steer towards a solution in which we actually give the user more fine-grained control over their stubbing logic vs removing it. E.x. As the ticket mentions, being able to control the mock data file at runtime by gaining access to inspect the query/body parameters and deciding which mock data file to use based on certain ones?